### ISSUE TITLE
Erroneous warning for loop name shadowing in separate match arms

### ISSUE THREAD

**Issue by insanitybit** [rust-lang/rust#30359]
http://is.gd/XAabFg

Despite the loop names not shadowing each other, as they exist within separate match arms, there's a warning for this code.

<anon>:13:13: 15:14 warning: label name `'l` shadows a label name that is already in scope

**Comment by utkarshkukreti** [comment]
Shorter test case:

```rust
fn main() {
    match Some(0) {
        Some(_) => 'a: loop { break },
        None    => 'a: loop { break }
    }
}
```
Warnings:

```
<anon>:4:20: 4:38 warning: label name `'a` shadows a label name that is already in scope
<anon>:4         None    => 'a: loop { break }
                            ^~~~~~~~~~~~~~~~~~
<anon>:3:20: 3:38 note: shadowed label `'a` declared here
<anon>:3         Some(_) => 'a: loop { break },
```

**Comment by petrochenkov** [comment]
This is intentional (at least currently). Labels in a function are supposed to be uniquely named regardless of scoping. The warning message could be better though.

cc https://github.com/rust-lang/rust/pull/24162

**Comment by GuillaumeGomez** [comment]
It seems very clear to me. In which way could it be improved?

**Comment by jonas-schievink** [comment]
Maybe "shadows another label already defined in this function"?

**Comment by GuillaumeGomez** [comment]
@jonas-schievink: That cannot be this because a scope is more precise than a function. Example:

```Rust
fn labels() {
    {
        'a: for i in 0..10 {}
    }
    {
        'a: for i in 0..10 {}
    }
}
```

Same function, but not same scopes. Do you understand better?

**Comment by jonas-schievink** [comment]
The linked PR #24162 intentionally makes duplicated labels emit a warning (even when they mark different scopes)

**Comment by GuillaumeGomez** [comment]
Hum... Well then, if you want to avoid this in a function scope, the better would then be to add a help message after the warning: "To keep your code easy to read an maintain, try to avoid using the same label name in a function scope".

**Comment by petrochenkov** [comment]
Duplicate of #24278
